### PR TITLE
Fix nested complexity fragment sorting

### LINKED ISSUE #3207: Very Deeply Nested (5+ levels?) Fragments Fail Based Upon Order of Fragments In Query
Hello again, I have a similar issue to the one before reported in #3191. The fix to that issue, #3192, resolved some of our queries, but not all. Thank you for getting that up so fast!

## Description
Using GraphQL DotNet 5.3.1 a query like the below fails (simplified example from a more complex project):

```
query fragmentTest
{
  carMake(rowId: 1)
  {
    name
    ...modelLinesOfferedByCarMake
  }
}

fragment modelLinesOfferedByCarMake on CarMake
{
  modelLines(first: 100)
  {
    edges
    {
      node
      {
        name
        modelYear
        ... segmentDetail
        ... modelsInModelLine 
      }
    }
  }
}

fragment modelsInModelLine on ModelLine
{
  models(first: 100)
  {
    edges
    {
      node
      {
        name
        ...carsInModel
      }
    }
  }
}


fragment carsInModel on Model
{
  cars(first: 100)
  {
    edges
    {
      node
      {
        rowId
        name
        msrp
        ... optionsOnCar
      }
    }
  }
}

fragment optionsOnCar on Car
{
  options(first: 10000)
  {
    edges
    {
      node
      {
        oemName
        ... contentsInOption
      }
    }
  }
}

fragment segmentDetail on ModelLine
{
  segment
  {
    name
    ...marketCategoryDetail
  }
}

fragment contentsInOption on Option
{
  optionContents(first: 100)
  {
    edges
    {
      node
      {
        ...contentDetail
      }
    }
  }
}


fragment contentDetail on OptionContent
{
  content
  {
    name
  }
}

fragment marketCategoryDetail on MarketSegment
{
  marketCategory
  {
    name
  }
}

```

With an error like 

> "System.Collections.Generic.KeyNotFoundException: The given key 'contentsInOption' was not present in the dictionary.\r\n   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)\r\n   at GraphQL.Validation.Complexity.ComplexityVisitor.VisitFragmentSpreadAsync(GraphQLFragmentSpread fragmentSpread, AnalysisContext context) in /_/src/GraphQL/Validation/Complexity/ComplexityVisitor.cs:line 85

The full error goes on a while, but it appears related to #3030 Nested Fragments and #3191 Deeply Nested Fragments.

However, when we re-arrange the fragments to move the segmentDetail fragment to the beginning or the end of the modelsInModelLine -> carsInModel -> optionsOnCar -> contentsInOption -> contentDetail chain of fragments, the query succeeds as expected. If we try a query that looks like this:

```
query fragmentTest
{
  carMake(rowId: 1)
  {
    name
    ...modelLinesOfferedByCarMake
  }
}

fragment modelLinesOfferedByCarMake on CarMake
{
  modelLines(first: 100)
  {
    edges
    {
      node
      {
        name
        modelYear
        ... segmentDetail
        ... modelsInModelLine 
      }
    }
  }
}

// No longer breaks the chain
 
fragment segmentDetail on ModelLine
{
  segment
  {
    name
    ...marketCategoryDetail
  }
}

fragment modelsInModelLine on ModelLine
{
  models(first: 100)
  {
    edges
    {
      node
      {
        name
        ...carsInModel
      }
    }
  }
}


fragment carsInModel on Model
{
  cars(first: 100)
  {
    edges
    {
      node
      {
        rowId
        name
        msrp
        ... optionsOnCar
      }
    }
  }
}

fragment optionsOnCar on Car
{
  options(first: 10000)
  {
    edges
    {
      node
      {
        oemName
        ... contentsInOption
      }
    }
  }
}

fragment contentsInOption on Option
{
  optionContents(first: 100)
  {
    edges
    {
      node
      {
        ...contentDetail
      }
    }
  }
}

fragment contentDetail on OptionContent
{
  content
  {
    name
  }
}

fragment marketCategoryDetail on MarketSegment
{
  marketCategory
  {
    name
  }
}
```

Everything works as expected, every time.

Clear steps describing how to reproduce the issue.

Create two independent nested fragment chains in a query, one at least 3 levels deep. Place one of fragments in the shorter chain between the fragments of the longer chain in the query. Execute the query.

### Expected Result

The query should execute regardless of the order of the fragments.

### Actual result

The order of the fragments matters, throwing an error from the subsystem if one fragment chain breaks another long fragment chain 

### Environment

OS: Windows 10
GraphQL DotNet Version: 5.3.1
Dotnet Core version: 6.0.3

### PR DESCRIPTION
**Pull request by Shane32**
Added failing test

Demonstrating that by swapping the order of two fragments, the test either passes or fails

[UPDATE]
Fixed by second commit.
5 4 2 1 3 - order before
5 3 2 4 1 - order after

Now should work for all cases.

### CODE REVIEW

**Comment by sungam3r** [comment]
5.4 ?

**Comment by sungam3r** [comment]
WOW, works fine locally, failed on CI due to `ComplexityAnalyzer.cs:line 71 - foreach on Keys`

**Review by sungam3r — APPROVED** [review]

File: src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs [hunk 1]
```diff
@@ -55,17 +58,94 @@ internal ComplexityResult Analyze(GraphQLDocument doc, double avgImpact, int max
...
+
+        private static GraphQLFragmentDefinition GetFirstFragmentWithoutPendingDependencies(Dictionary<GraphQLFragmentDefinition, List<GraphQLFragmentDefinition>?> dependencies)
+        {
+            foreach (var item in dependencies)
+            {
+                if (item.Value == null)
+                    return item.Key;
+            }
+
+            throw new InvalidOperationException("Fragments dependency cycle detected!");
```
Comment (sungam3r) [inline src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs:106]: #3178

**Comment by Shane32** [comment]
> WOW, works fine locally, failed on CI due to `ComplexityAnalyzer.cs:line 71 - foreach on Keys`

I think it might be due to different frameworks.

**Comment by Shane32** [comment]
Looks good to me.  I'll merge and make a release.

**Comment by Shane32** [comment]
This is the only PR merged since 5.3.1 so this will be 5.3.2 as it is a bugfix

**Pull request merged**